paasta_tools/cli/cmds/status.py

```diff
@@ -1277,41 +1292,78 @@ def get_version_table_entry(
+                    unhealthy_replicas, service, instance, verbose
+                )
                 for line in replica_table:
                     entry.append(f"      {line}")
         else:
-            replica_table = create_replica_table(replica_states)
+            replica_table = create_replica_table(
+                replica_states, service, instance, verbose
+            )
             for line in replica_table:
                 entry.append(f"    {line}")
     return entry
 
 
 class ReplicaState(Enum):
-    RUNNING = "Healthy"
-    UNSCHEDULED = "Unscheduled"
-    STARTING = "Starting"
-    WARMING_UP = "Warming Up"
-    TERMINATING = "Terminating"
-    HEALTHCHECK_FAILING = "Healthcheck Failing"
-    UNREACHABLE = "Unreachable"
-    UNHEALTHY = "Unhealthy"
-    UNKNOWN = "Unknown"
+    # Order will be preserved in count summary
+    RUNNING = "Healthy", PaastaColors.green
+    UNREACHABLE = "Unreachable", PaastaColors.red
+    NOT_RUNNING = "Not Running", PaastaColors.red
+    WARNING = "Warning", PaastaColors.yellow
+    UNSCHEDULED = "Unscheduled", PaastaColors.yellow
+    STARTING = "Starting", PaastaColors.yellow
+    WARMING_UP = "Warming Up", PaastaColors.cyan
+    TERMINATING = "Terminating", PaastaColors.cyan
+    UNKNOWN = "Unknown", PaastaColors.yellow
 
     def is_unhealthy(self):
-        return self in ReplicaState.unhealthy_states()
+        return self.name in ["UNREACHABLE", "NOT_RUNNING"]
+
+    @property
+    def color(self) -> Callable:
+        return self.value[1]
 
     @property
-    def color(self):
-        if self.is_unhealthy():
-            return PaastaColors.red
-        elif self.name == "RUNNING":
-            return PaastaColors.green
-        return PaastaColors.yellow
+    def formatted_message(self):
+        return self.value[1](self.value[0])
 
-    @classmethod
-    def unhealthy_states(cls):
-        return [cls.UNHEALTHY, cls.UNREACHABLE, cls.HEALTHCHECK_FAILING]
+    @property
+    def message(self):
+        return self.value[0]
+
+
+def recent_liveness_failure(pod: KubernetesPodV2) -> bool:
+    if not pod.events:
+        return False
+    return any(
+        [evt for evt in pod.events if "Liveness probe failed" in evt.get("message", "")]
+    )
+
+
+def recent_container_restart(
+    container: KubernetesContainerV2, time_window: int = 900
+) -> bool:
+    min_timestamp = datetime.now().timestamp() - time_window
+    if (
+        container.restart_count > 0
+        and container.last_state == "terminated"
+        and container.last_timestamp > min_timestamp
+    ):
+        return True
+    return False
+
+
+def get_main_container(pod: KubernetesPodV2) -> Optional[KubernetesContainerV2]:
+    return next(
+        (
+            c
+            for c in pod.containers
+            if c.name not in kubernetes_tools.SIDECAR_CONTAINER_NAMES
+        ),
+        None,
+    )
 
 
 def get_replica_state(pod: KubernetesPodV2) -> ReplicaState:
```

Comment:
I kind of think we should consider merging the logic that assigns state in this method with the logic that adds tips in `create_replica_table` since the latter uses the state determination in the former.  I think that the state and tip stuff is the most complex bit of "business logic" here, so it'd be good to try to keep it together and make it as easy as possible to reason about.

That said, probably too complicated for this PR -- just want to throw it out there.

Reply:
Yeah I do agree, we're doing a lot of the same work in both, but I kind of dug myself into this pattern already.  For now I'll punt on that and I can ticket it separately.